### PR TITLE
Added missing tests

### LINKED ISSUE #765: Add missing tests for UnusedReturnValueAnalyzer
Not all the [listed methods](https://github.com/FakeItEasy/FakeItEasy/blob/508987ae8aa5599d29ea09668fab744a76d2d4e5/src/FakeItEasy.Analyzer/UnusedReturnValueAnalyzer.cs#L72-L82) are covered.

### PR DESCRIPTION
**Pull request by medmondson**
Add the four missing tests for #765 

fixes #765

### CODE REVIEW

**Comment by blairconrad** [comment]
Thanks for the contribution, @medmondson! We appreciate it.
I'll take a gander over the next day or so!

**Comment by medmondson** [comment]
Thanks for the review @blairconrad.

Don't worry I'll pick up all the points you mentioned, I must admit I did just use the original list without checking further.

My VB is a bit rusty but I'm sure I can figure it out.  All these updates are improving my skills with git so I appreciate the extra work.

**Comment by blairconrad** [comment]
> I must admit I did just use the original list without checking further.

Well, it's entirely possible that I did a poor job of making it at the time! Or maybe something changed. I'm too lazy to look.

> My VB is a bit rusty but I'm sure I can figure it out.

Yeah, I have no VB since... 5? And that was about 16 years ago. I just copy the examples. I don't think it should be _too_ tricky.

> All these updates are improving my skills with git so I appreciate the extra work.

Then we both win! I'm glad you bring it up. I forgot to mention that the changes here and that you'll be making in the next step are simple enough that it doesn't matter to me whether you amend your first commit, or make new commits and squash later. Whatever easiest for you. Or exercises the skill you want to develop, I suppose!

Thanks for your continued efforts, @medmondson.

**Comment by medmondson** [comment]
Okay @blairconrad I've ticked off the remaining tasks and would really appreciate if you could eyeball the changes. Any feedback appreciated.

Once we're happy I'll squash (and rebase if necessary) to get this PR into a mergeable state.

**Comment by blairconrad** [comment]
Thanks for the update, @medmondson. I'll take a look soonish!

**Comment by blairconrad** [comment]
@medmondson, I'm too tired/lazy to actually look at code tonight, but I did run the C# version of your tests and found that
"FakeItEasy.Configuration.IWhereConfiguration`1.Where" and "FakeItEasy.A.CallTo" remain uncovered while

"FakeItEasy.A.CallTo`1" is covered by both
* `Diagnostic_Should_Be_Triggered_When_Call_Specification_Is_Not_Used` and
* `Diagnostic_Should_Be_Triggered_When_Call_Specification_Made_In_Global_Scope_Is_Not_Used`

"FakeItEasy.Configuration.IPropertySetterAnyValueConfiguration`1.To"  is covered by both
* `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_CallToSet_To_Value` and
* `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_CallToSet_To_Expression`

"FakeItEasy.WhereConfigurationExtensions.Where`1"  is covered by both
* `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_Where` and
* `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_Where_With_No_Return_Type`

I'm happy to go back and read things, but think it may be premature. At the very least, I expect to find the same situation for the VB tests. How would you like to proceed?

**Comment by medmondson** [comment]
Apologies, I was certain I'd covered them off, are you using a tool to discover coverage? I would say I probably spent the majority of my time 'discovering' the way the library is used for these methods so they're very much new to me.

I'll review them all again, removing the duplicates and modifying to cover those missed.

You're right the VB tests will be the same

**Comment by blairconrad** [comment]
@medmondson, I don't have a fancy tool. I've been thinking that it may be worthwhile developing something to help with this, but haven't thought of what we could do.
Here's my approach, though. It's a little clunky:

1. add a breakpoint to `DiagnosticVerifier.VerifyCSharpDiagnostic` (and I suppose `DiagnosticVerifier.VerifyVisualBasicDiagnostic` at some point, but I've not done yet) whose action is to print its `CALLER`. This will locate the test being run:
image
2. add a breakpoint to `UnusedReturnValueAnalyzer.AnalyzeCall`, whose action is to print `methodFullName`. This will identify the "callSpecMemberName" that's hit.
image

Then I just read the Output. and compare against the full list of `callSpecMemberNames` to see what's missing, and also to see who was responsible for what.

Give it a try. It may help. Or if you develop something more sophisticated, that'd be cool too.

Thanks for sticking with this!

**Comment by medmondson** [comment]
@blairconrad - Amazing tips, thank you.

Having just ran this through it looks as though:

FakeItEasy.A.CallTo is covered by

- `Diagnostic_Should_Have_The_Correct_Call_Description_For_Call_To_With_No_Expression`

with the following correctly identified as duplicates:

- FakeItEasy.A.CallTo`1
- FakeItEasy.Configuration.IPropertySetterAnyValueConfiguration`1.To
- FakeItEasy.WhereConfigurationExtensions.Where`1

Although they don't appear to be completely identical, for example

`Diagnostic_Should_Be_Triggered_When_Call_Specification_Is_Not_Used()`

and

`Diagnostic_Should_Be_Triggered_When_Call_Specification_Made_In_Global_Scope_Is_Not_Used()`

both test ```FakeItEasy.A.CallTo`1``` but are checking for a difference when used with the `using static` construct. Is this not a valid, separate test?

In addition I'm only responsible for the last of these duplicates in `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_Where_With_No_Return_Type` 

I'm happy to take your lead on if the above are in fact duplicate (and one from each needs removed) or if they are seperate scenarios being tested.

Either way I can see we're definitely missing a test for FakeItEasy.Configuration.IWhereConfiguration`1.Where so I'll get that written and I'll await your comments on the above.

Once again thanks for the tips and continued feedback, it is thoroughly enjoyable to contribute.

**Comment by blairconrad** [comment]
> `Diagnostic_Should_Be_Triggered_When_Call_Specification_Is_Not_Used()` and `Diagnostic_Should_Be_Triggered_When_Call_Specification_Made_In_Global_Scope_Is_Not_Used()`
both test FakeItEasy.A.CallTo`1 but are checking for a difference when used with the using static construct. Is this not a valid, separate test?

@medmondson, you're absolutely right. That was sloppy of me. I really should've looked closely at the names of the tests, both to see which were valid duplicates and to see that they were pre-existing. I'd thought to point out which tests covered duplicate cases as a diagnostic aid, but I really should've gone deepr.

Even having `Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_Where_With_No_Return_Type` isn't a _bad_ thing.

> `FakeItEasy.A.CallTo` is covered by `Diagnostic_Should_Have_The_Correct_Call_Description_For_Call_To_With_No_Expression`

So it is! I do not know how missed it. Sorry.


Yeah, I think all we are missing is tests for ``FakeItEasy.Configuration.IWhereConfiguration`1.Where``.

Thanks for your patience!

**Comment by medmondson** [comment]
@blairconrad - There we go, I've made sure I've covered the remaining case.

As it's getting late now (and I've likely made mistakes) I'd be grateful if you could check over what I have so far.  If everything is good I'll sort out getting the branch into a mergeable state.

Many thanks

**Comment by blairconrad** [comment]
@medmondson, I took a very quick eyeball at the new commit, and I think it looked good, but paying work interferes and I didn't have much time. I'll get back to you in a day or so, if the fates align.
As it is, there's no huge rush, since we just released 3.0.0-rc001 last week, and while these changes will not affect the generated packages, it might be cleaner to wait until we release 3.0.0 before we merge to master.

Thanks!

File: tests/FakeItEasy.Analyzer.CSharp.Tests/UnusedReturnValueAnalyzerTests.cs [hunk 1]
```diff
@@ -256,6 +347,67 @@ interface IFoo { int Bar(); }
...
+                    Message =
+                        "Unused call specification 'A.CallTo(fake).Where(call => true, output => new object())'; did you forget to configure or assert the call?",
+                    Severity = DiagnosticSeverity.Error,
+                    Locations = new[] { new DiagnosticResultLocation("Test0.cs", 9, 13) }
+                });
+        }
+
+        [Fact]
+        [UsingCulture("en-US")] // so that the message is in the expected language regardless of the OS language
+        public void Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_WhereArgumentsMatch_With_No_Return_Specified()
```
Comment (blairconrad) [inline tests/FakeItEasy.Analyzer.CSharp.Tests/UnusedReturnValueAnalyzerTests.cs]: "WhenArgumentsMatch"

File: tests/FakeItEasy.Analyzer.VisualBasic.Tests/UnusedReturnValueAnalyzerTests.cs [hunk 2]
```diff
@@ -249,6 +342,66 @@ End Namespace
...
+                    Id = DiagnosticDefinitions.UnusedCallSpecification.Id,
+                    Message =
+                        "Unused call specification 'A.CallTo(foo).Where(Function() True, New Object())'; did you forget to configure or assert the call?",
+                    Severity = DiagnosticSeverity.Error,
+                    Locations = new[] { new DiagnosticResultLocation("Test0.vb", 6, 13) }
+                });
+        }
+
+        [Fact]
+        public void Diagnostic_Should_Have_The_Correct_Call_Description_If_Triggered_On_WhereArgumentsMatch_With_No_Return_Specified()
```
Comment (blairconrad) [inline tests/FakeItEasy.Analyzer.VisualBasic.Tests/UnusedReturnValueAnalyzerTests.cs]: Ditto, "WhenArgumentsMatch"

**Comment by medmondson** [comment]
@blairconrad thanks, that's now done.

Not wanting to be to be a burden but could you please take a look at my branch?  It appears my attempted rebase has duplicated all of the commits and I'm not sure what my next step should be to fix it, or even how I got there in the first place.

It's okay if you're too busy, I'll figure it out eventually.

**Comment by blairconrad** [comment]
Hey, @medmondson. Never a burden. I'm looking at the graph and I'm not sure what you did.
If I had to make up series of steps it would be:

* push the typo commit
* rebase the local branch onto upstream/master
* merge the origin branch

but all of that seems pretty unlikely. I do believe it's fixable. I'm not sure what you're using to view the commits, but in GitKraken, which I find lays things out pretty clearly, I see
image

You can see two of branches that have your 6 commits. One was taken of what was then upstream/master about 2 weeks ago: 
"Merge pull request #974 from thomaslevesque/upgrade-castle-core-4.0.0" 
at commit aef21af4a5b759198d0a3613c04203e80df7e787.

The next comes right off of current upstream/master:
"Merge pull request #977 from thomaslevesque/fix-nuspec-castle-version"
at commit 7faa7e8a9de48c4bf01364fe4caed890d10c80a9.

Me, I think you want your branch to be the last commit on this second branch: 
"Fix typos" at commit 32562a462fb58b62abb9d1b9c5cd23f98017831a.

I think you can accomplish the fix with a 

```
git reset --hard 32562a462fb58b62abb9d1b9c5cd23f98017831a
```

(when you're in a command prompt with `MissingTests` checked out).

Then a `git push origin HEAD --force` should fix you right up.

**Comment by medmondson** [comment]
@blairconrad - After much reading (and a dash of frustration) I think I'm there.  Please check. It should be a nicely squashed commit straight off the main.

Many thanks for your continued help!

**Comment by blairconrad** [comment]
@medmondson, I'm glad it worked. It's just as you say, a nicely squashed commit straight off master.
I'm sorry you had much reading and some frustration. I hope there wasn't a deficiency in my explanation that caused you the frustration.

@adamralph, @thomaslevesque, I'm happy with the changes, and am happy if we merge. I'll leave open for the weekend in case you want to take a look, though. No pressure.

**Comment by medmondson** [comment]
@blairconrad, thank you. Not at all, your explanation was perfect and commands fixed it straight away. I was simply trying to understand how I'd got into that mess as it was the first time I used gitkraken for the rebase.

It's all good learning though.

**Comment by thomaslevesque** [comment]
I only made a cursory review, but everything looks fine. Merging now.

Thanks @medmondson!

**Pull request merged**